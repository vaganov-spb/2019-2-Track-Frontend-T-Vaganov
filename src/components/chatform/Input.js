import React from 'react';
import { connect } from 'react-redux';
import { setTime } from '../../utils';
import { getMessagesSuccess, changeMessageInputValue, clearMessageInputValue } from '../../actions';
import formStyles from '../../styles/FormInput.module.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import 'font-awesome/css/font-awesome.min.css';

function Input (props){
	function onKeyPress(event) {
		if (event.key === 'Enter') {
			event.preventDefault();
			if(props.canSend) {
				if(props.chatId !== 'Group Chat'){
					props.addMessages(
						[{message: `${props.currentText}`, type: 'text', time: `${setTime()}`}],
						props.chatId
					);
				}  else {
					fetch('http://localhost:8000/chats/1/newmessage/',{
						method: 'POST',
						headers: {'Content-Type': 'application/json'},
						body: JSON.stringify({content: `${props.currentText}`, chatId:21})
					});
				}
				props.clearInput(props.chatId);
			}
		}
	}

	function handleChange(event) {
		if(event.target.value) {
			props.changeInput(event.target.value, props.chatId);
		} else {
			props.clearInput(props.chatId);
		}
	};

	return (
		<input
			className={formStyles.form}
			type="text"
			value={props.currentText}
			placeholder="Cooбщение"
			onChange={handleChange}
			onKeyPress={onKeyPress}
		/>
	);
}


const mapStateToProps = (state, ownProps) => {
	return {
		currentText: state.currentInputText[ownProps.chatId].text,
		canSend: state.currentInputText[ownProps.chatId].flag,
	};
};

const mapDispatchToProps = (dispatch) => ({
	addMessages: (message, chatId) => dispatch(getMessagesSuccess(message, chatId)),
	changeInput: (text, chatId) => dispatch(changeMessageInputValue(text, chatId)),
	clearInput: (chatId) => dispatch(clearMessageInputValue(chatId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Input);