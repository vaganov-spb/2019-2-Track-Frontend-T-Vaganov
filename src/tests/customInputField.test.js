/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Input from '../components/chatform/customInputField';

const mockStore = configureStore();

describe('customInputFields renders and act correctly', () => {
	let store;
	let component;
	let props;

	beforeEach(() => {
		props = {
			chatId: 1,
		};

		store = mockStore({
			currentInputText: {
				1: {
					text: 'test_text',
					flag: 'true',
				}
			},
		});
		store.dispatch = jest.fn();
		component = renderer.create(
			<Provider store={store}>
				<MemoryRouter>
					<Input { ...props }/>
				</MemoryRouter>  
			</Provider>
		);
	});

	it ('should render normally', () => {
		expect(component.toJSON()).toMatchSnapshot();
	});
   
	it ('should work correctly onChange 1', () => {
		const value = component.root.findByProps({className: 'form'});
		const event = {
			preventDefault() {},
			target: { value: '' }
		};

		value.props.onChange(event);
		expect(store.dispatch).toHaveBeenCalledTimes(1);
		expect(store.dispatch).toHaveBeenCalledWith({'chat': 1, 'type': 'CLEAR_INPUT_VAL'});
	});

	it ('should work correctly onChange 2', () => {
		const value = component.root.findByProps({className: 'form'});
		const event = {
			preventDefault() {},
			target: { value: 'test_text' }
		};

		value.props.onChange(event);
		expect(store.dispatch).toHaveBeenCalledTimes(1);
		expect(store.dispatch).toHaveBeenCalledWith({'chat': 1, 'type': 'CHANGE_INPUT_VAL', 'payload': 'test_text'});
	});
    
	it ('should work correctly onKeyPress 1', () => {
		const value = component.root.findByProps({className: 'form'});
		const event = {
			preventDefault() {},
			key: 'Enter',
		};

		value.props.onKeyPress(event);
		expect(store.dispatch).toHaveBeenCalledTimes(2);
		expect(store.dispatch).toHaveBeenNthCalledWith(1, {
			'chat': 1,
			'payload': [
				{
					'message': 'test_text',
					'time': expect.anything(),
					'type': 'text',
				},
			],
			'type': 'LOAD_MESSAGES_SUCCESS',
		}); 
		expect(store.dispatch).toHaveBeenNthCalledWith(2, { 'chat': 1, 'type': 'CLEAR_INPUT_VAL' });
	});
});