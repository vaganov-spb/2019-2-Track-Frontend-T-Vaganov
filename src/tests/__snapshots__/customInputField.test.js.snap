// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customInputFields renders and act correctly should render normally 1`] = `
<input
  className="form"
  onChange={[Function]}
  onKeyPress={[Function]}
  placeholder="Cooбщение"
  type="text"
  value="test_text"
/>
`;
